python/cuml/experimental/cluster/robust_single_linkage.pyx

```diff
@@ -0,0 +1,256 @@
+                                     int *mst_dst,
+                                     float *mst_weights)
+        int get_n_leaves()
+        int get_n_clusters()
+
+    cdef cppclass RobustSingleLinkageParams:
+        int k
+        int min_samples
+        int min_cluster_size
+
+        float alpha
+
+        float cluster_selection_epsilon,
+
+        bool allow_single_cluster,
+
+
+cdef extern from "cuml/cluster/hdbscan.hpp" namespace "ML":
+
+    void robust_single_linkage(const handle_t &handle,
+                               const float *X,
+                               size_t m,
+                               size_t n,
+                               DistanceType metric,
+                               RobustSingleLinkageParams &params,
+                               robust_single_linkage_output &out)
+
+
+def delete_output(obj):
+    cdef robust_single_linkage_output *output
+    if hasattr(obj, "hdbscan_output_"):
+        output = <robust_single_linkage_output*>\
+                  <uintptr_t> obj.rsl_output_
+        del output
+        del obj.rsl_output_
+
+
+_metrics_mapping = {
+    'l1': DistanceType.L1,
+    'cityblock': DistanceType.L1,
+    'manhattan': DistanceType.L1,
+    'l2': DistanceType.L2SqrtExpanded,
+    'euclidean': DistanceType.L2SqrtExpanded,
+    'cosine': DistanceType.CosineExpanded
+}
+
+
+class RobustSingleLinkage(Base, ClusterMixin, CMajorInputTagMixin):
+    """Perform robust single linkage clustering from a vector array
+    or distance matrix.
+    Robust single linkage is a modified version of single linkage that
+    attempts to be more robust to noise. Specifically the goal is to
+    more accurately approximate the level set tree of the unknown
+    probability density function from which the sample data has
+    been drawn.
+
+    Parameters
+    ----------
+    X : array of shape (n_samples, n_features), or \
+            array of shape (n_samples, n_samples)
+        A feature array
+    cut : float
+        The reachability distance value to cut the cluster heirarchy at
+        to derive a flat cluster labelling.
+    k : int, optional (default=5)
+        Reachability distances will be computed with regard to the `k`
+        nearest neighbors.
+    alpha : float, optional (default=np.sqrt(2))
+        Distance scaling for reachability distance computation. Reachability
+        distance is computed as
+        max { core_k(a), core_k(b), 1/alpha * d(a,b) }.
+    gamma : int, optional (default=5)
+        Ignore any clusters in the flat clustering with size less than gamma,
+        and declare points in such clusters as noise points.
+    metric : string, optional (default='euclidean')
+        The metric to use when calculating distance between instances in a
+        feature array.
+    metric_params : dict, option (default={})
+        Keyword parameter arguments for calling the metric (for example
+        the p values if using the minkowski metric).
```

Comment:
Any particular reason to not follow the same API here and in the HDBSCAN class?

Reply:
This is just a direct port of the `RobustSingleLinkage` in the HDBSCAN codebase so I was just following what they did